coordinator/src/tributary/scanner.rs

```diff
@@ -114,12 +117,38 @@ pub(crate) async fn handle_new_blocks<
   processors: &Pro,
   publish_serai_tx: PST,
   spec: &TributarySpec,
-  tributary: &TributaryReader<D, Transaction>,
+  tributary: &Arc<Tributary<D, Transaction, P>>,
+  tributary_reader: &TributaryReader<D, Transaction>,
 ) {
-  let genesis = tributary.genesis();
+  let genesis = tributary_reader.genesis();
   let mut last_block = db.last_block(genesis);
-  while let Some(next) = tributary.block_after(&last_block) {
-    let block = tributary.block(&next).unwrap();
+  while let Some(next) = tributary_reader.block_after(&last_block) {
+    let block = tributary_reader.block(&next).unwrap();
+
+    // check whether we have the provided txs provided to us from processor.
+    // As long as we don't have all the provided txs, we just delay the scanning
+    // of the block until we have all of them.
+    let provided_transactions = tributary.provided_transactions().await;
+    for tx in &block.transactions {
+      if let TransactionKind::Provided(_) = tx.kind() {
+        let tx_hash = tx.hash();
+        if !provided_transactions.chunks_exact(32).any(|hash| {
+          let h: [u8; 32] = hash.try_into().unwrap();
+          h == tx_hash
+        }) {
+          return;
+        }
+      }
+    }
```

Comment:
This is an O(n^2) algorithm which breaks safety. Provided are expected to be ordered, yet this disregards order.

IF we were going to have this unsafe algorithm, though no, we do need to keep the ordering checks, we could at least make it O(n) with a HashSet...